packages/expo-facebook/android/src/main/java/expo/modules/facebook/FacebookModule.kt

```diff
@@ -0,0 +1,342 @@
+    promise.resolve(null)
+  }
+
+  @ExpoMethod
+  fun logEventAsync(eventName: String?, valueToSum: Double, parameters: ReadableArguments?, promise: Promise) {
+    appEventLogger!!.logEvent(eventName, valueToSum, bundleWithNullValuesAsStrings(parameters))
+    promise.resolve(null)
+  }
+
+  @ExpoMethod
+  fun logPurchaseAsync(
+    purchaseAmount: Double,
+    currencyCode: String?,
+    parameters: ReadableArguments?,
+    promise: Promise
+  ) {
+    try {
+      appEventLogger!!.logPurchase(
+        BigDecimal.valueOf(purchaseAmount),
+        Currency.getInstance(currencyCode),
+        bundleWithNullValuesAsStrings(parameters)
+      )
+      promise.resolve(null)
+    } catch (e: Exception) {
+      promise.reject("ERR_FACEBOOK_APP_EVENT_LOGGER", "appEventLogger is not initialized", e)
+    }
+  }
+
+  @ExpoMethod
+  fun logPushNotificationOpenAsync(campaign: String?, promise: Promise) {
+    // the Android FBSDK expects the fb_push_payload to be a JSON string
+    val payload = Bundle()
+    payload.putString(PUSH_PAYLOAD_KEY, String.format("{\"%s\" : \"%s\"}", PUSH_PAYLOAD_CAMPAIGN_KEY, campaign))
+    try {
+      appEventLogger!!.logPushNotificationOpen(payload)
+      promise.resolve(null)
+    } catch (e: Exception) {
+      promise.reject("ERR_FACEBOOK_APP_EVENT_LOGGER", "appEventLogger is not initialized", e)
+    }
+  }
+
+  @ExpoMethod
+  fun setUserIDAsync(userID: String?, promise: Promise) {
+    AppEventsLogger.setUserID(userID)
+    promise.resolve(null)
+  }
+
+  @ExpoMethod
+  fun getUserIDAsync(promise: Promise) {
+    promise.resolve(AppEventsLogger.getUserID())
+  }
+
+  @ExpoMethod
+  fun getAnonymousIDAsync(promise: Promise) {
+    try {
+      promise.resolve(AppEventsLogger.getAnonymousAppDeviceGUID(context))
+    } catch (e: Exception) {
+      promise.reject("ERR_FACEBOOK_ANONYMOUS_ID", "Can not get anonymousID", e)
+    }
+  }
+
+  @ExpoMethod
+  fun getAdvertiserIDAsync(promise: Promise) {
+    try {
+      promise.resolve(attributionIdentifiers!!.androidAdvertiserId)
+    } catch (e: Exception) {
+      promise.reject("ERR_FACEBOOK_ADVERTISER_ID", "Can not get advertiserID", e)
+    }
+  }
+
+  @ExpoMethod
+  fun getAttributionIDAsync(promise: Promise) {
+    try {
+      promise.resolve(attributionIdentifiers!!.attributionId)
+    } catch (e: Exception) {
+      promise.reject("ERR_FACEBOOK_ADVERTISER_ID", "Can not get attributionID", e)
+    }
+  }
+
+  @ExpoMethod
+  fun setUserDataAsync(userData: ReadableArguments, promise: Promise) {
+    AppEventsLogger.setUserData(
+      userData.getString("email"),
+      userData.getString("firstName"),
+      userData.getString("lastName"),
+      userData.getString("phone"),
+      userData.getString("dateOfBirth"),
+      userData.getString("gender"),
+      userData.getString("city"),
+      userData.getString("state"),
+      userData.getString("zip"),
+      userData.getString("country")
+    )
```

Comment:
I don't know if this will make this more readable or rather less 🤔
```suggestion
    with(userData) {
      AppEventsLogger.setUserData(
        getString("email"),
        getString("firstName"),
        getString("lastName"),
        getString("phone"),
        getString("dateOfBirth"),
        getString("gender"),
        getString("city"),
        getString("state"),
        getString("zip"),
        getString("country")
      )
    }
```